File: common/src/main/java/dev/mayaqq/estrogen/client/registry/blockRenderers/dreamBlock/texture/advanced/Goober.java
package dev.mayaqq.estrogen.client.registry.blockRenderers.dreamBlock.texture.advanced;

import com.mojang.blaze3d.platform.NativeImage;
import com.mojang.blaze3d.systems.RenderSystem;
import net.minecraft.util.FastColor;
import net.minecraft.util.RandomSource;
import net.minecraft.util.random.SimpleWeightedRandomList;
import net.minecraft.util.random.Weight;
import net.minecraft.util.random.WeightedEntry;
import net.minecraft.util.random.WeightedRandomList;
import org.joml.Math;

import java.util.List;

public class Goober {

    static final SimpleWeightedRandomList<Integer> TRANSPARENCY = new SimpleWeightedRandomList.Builder<Integer>().add(0, 5).add(1, 2).add(2, 1).build();

    private final int x;
    private final int y;
    private final Color color;
    private final Style style;
    private final int frameTick;
    private final int transparencyLevel;
    private int currentFrame;

    public Goober(int x, int y, Color color, Style style, int frameTick, int beginFrame, int transparencyLevel) {
        this.x = x;
        this.y = y;
        this.color = color;
        this.style = style;
        this.frameTick = frameTick;
        this.currentFrame = beginFrame;
        this.transparencyLevel = transparencyLevel;
    }

    public boolean tooClose(int x, int y) {
        int difX = Math.abs(this.x - x);
        int difY = Math.abs(this.y - y);
        return (difX < 8 && difY < 8 || ((difX < 6 || difY < 6) && style != Style.PIXEL));
    }

    public void draw(NativeImage pixels) {
        int col = color.color;
        switch (transparencyLevel) {
            case 1: {
                col = FastColor.ARGB32.multiply(col, 0xFEFFFFFF);
                col = FastColor.ARGB32.multiply(col, 0xFEFFFFFF);
            }
            case 2: {
                col = FastColor.ARGB32.multiply(col, 0xFEFFFFFF);
                col = FastColor.ARGB32.multiply(col, 0xFEFFFFFF);
                col = FastColor.ARGB32.multiply(col, 0xFEFFFFFF);
            }
        }
        style.draw(pixels, x, y, col, currentFrame);
    }

    public void tickAnimation(int tick, Runnable redraw) {
        if(tick == frameTick) {
            currentFrame++;
            if(currentFrame == style.frameCount()) {
                currentFrame = 0;
            }
            RenderSystem.recordRenderCall(redraw::run);
        }
    }

    public enum Color {

        YELLOW(rgb(255, 255, 0)),
        CYAN(rgb(0, 241, 254)),
        PURPLE(rgb(106, 106, 198)),
        MAGENTA(rgb(255, 71, 231)),
        GREEN1(rgb(40, 125, 77)),
        GREEN2(rgb(40, 198, 53));

        final int color;
        Color(int color) {
            this.color = color;
        }

        public static int rgb(int r, int g, int b) {
            return FastColor.ARGB32.color(255, b, g, r); // abgr moment
        }
    }

    public enum Style implements WeightedEntry {
        PIXEL(2, List.of(NativeImage::setPixelRGBA)), // Pixel has a lower weight because any node drawn on a border becomes a pixel
        STAR(5, List.of((image, x, y, color) -> {
                image.setPixelRGBA(x + 1, y, color);
                image.setPixelRGBA(x, y + 1, color);
                image.setPixelRGBA(x - 1, y, color);
                image.setPixelRGBA(x, y - 1, color);
            }
        )),
        THINGY(2, List.of((pixels, x, y, col) -> {
            pixels.setPixelRGBA(x + 1, y, col);
            pixels.setPixelRGBA(x, y + 1, col);
            pixels.setPixelRGBA(x - 1, y, col);
            pixels.setPixelRGBA(x, y - 1, col);

            int transCol = FastColor.ARGB32.multiply(col, 0xFE999999);
            pixels.setPixelRGBA(x + 1, y + 1, transCol);
            pixels.setPixelRGBA(x - 1, y - 1, transCol);
        }
        )),
        STAR_ANIMATED(3, List.of(
            (image, x, y, color) -> {
                image.setPixelRGBA(x + 1, y, color);
                image.setPixelRGBA(x, y + 1, color);
                image.setPixelRGBA(x - 1, y, color);
                image.setPixelRGBA(x, y - 1, color);
            },
            (image, x, y, color) -> {
                image.setPixelRGBA(x + 1, y, color);
                image.setPixelRGBA(x, y + 1, color);
                image.setPixelRGBA(x - 1, y, color);
                image.setPixelRGBA(x, y - 1, color);
                image.setPixelRGBA(x + 2, y, color);
                image.setPixelRGBA(x - 2, y,  color);
                image.setPixelRGBA(x, y + 2, color);
                image.setPixelRGBA(x, y - 2, color);

                int transCol = FastColor.ARGB32.multiply(color, 0xFE999999);
                image.setPixelRGBA(x + 1, y + 1, transCol);
                image.setPixelRGBA(x - 1, y - 1, transCol);
                image.setPixelRGBA(x - 1, y + 1, transCol);
                image.setPixelRGBA(x + 1, y - 1, transCol);
            }
        ));

        final Weight weight;
        final List<DrawFunction> frames;

        Style(int weight, List<DrawFunction> frames) {
            this.weight = Weight.of(weight);
            this.frames = frames;
        }

        void draw(NativeImage image, int x, int y, int color, int frame) {
            frames.get(frame).draw(image, x, y, color);
        }

        public int frameCount() {
            return frames.size();
        }

        public boolean hasAnimation() {
            return frames.size() > 1;
        }

        @Override
        public Weight getWeight() {
            return this.weight;
        }

        private static final WeightedRandomList<Style> weightedRandomList = WeightedRandomList.create(values());

        public static Style weighted(RandomSource rng) {
            return weightedRandomList.getRandom(rng).get();
        }

        interface DrawFunction {
            public void draw(NativeImage image, int x, int y, int color);
        }
    }
}

File: src/main/kotlin/dev/mayaqq/estrogen/client/content/blockRenderers/dreamBlock/texture/Goober.kt
package dev.mayaqq.estrogen.client.content.blockRenderers.dreamBlock.texture

import com.mojang.blaze3d.platform.NativeImage
import invoke.kitty.kritter.utils.color.Color
import invoke.kitty.kritter.utils.color.Yellow
import invoke.kitty.kritter.utils.color.rgb
import invoke.kitty.kritter.utils.color.toColor
import net.minecraft.util.FastColor
import net.minecraft.util.RandomSource
import net.minecraft.util.random.SimpleWeightedRandomList
import net.minecraft.util.random.Weight
import net.minecraft.util.random.WeightedEntry
import net.minecraft.util.random.WeightedRandomList
import org.joml.Math

class Goober(
    private val x: Int,
    private val y: Int,
    private val color: Colors,
    private val style: Style,
    private val frameTick: Int,
    private var currentFrame: Int,
    private val transparencyLevel: Int
) {
    fun tooClose(x: Int, y: Int): Boolean {
        val difX = Math.abs(this.x - x)
        val difY = Math.abs(this.y - y)
        if (style == Style.PIXEL) return (difX < 2 && difY < 2)
        return (difX < 8 && difY < 8 || difX < 6 || difY < 6)
    }

    fun draw(pixels: NativeImage) {
        var col = color.color
        when (transparencyLevel) {
            1 -> {
                col *= 0xFEFFFFFFu.toColor()
                col *= 0xFEFFFFFFu.toColor()
                col *= 0xFEFFFFFFu.toColor()
                col *= 0xFEFFFFFFu.toColor()
                col *= 0xFEFFFFFFu.toColor()
            }

            2 -> {
                col *= 0xFEFFFFFFu.toColor()
                col *= 0xFEFFFFFFu.toColor()
                col *= 0xFEFFFFFFu.toColor()
            }
        }
        style.draw(pixels, x, y, col, currentFrame)
    }

    fun tickAnimation(tick: Int): Boolean {
        if (tick == frameTick) {
            currentFrame++
            if (currentFrame == style.frameCount()) {
                currentFrame = 0
            }
            return true
        }
        return false
    }

    enum class Colors(val color: Color) {
        YELLOW(Yellow),
        CYAN(rgb(0, 241, 254)),
        PURPLE(rgb(126, 126, 218)),
        MAGENTA(rgb(255, 71, 231)),
        GREEN1(rgb(60, 145, 97)),
        GREEN2(rgb(40, 198, 53));
    }

    enum class Style(weight: Int, val frames: List<DrawFunction>) : WeightedEntry {
        PIXEL(
            2,
            listOf(
                DrawFunction { obj: NativeImage, x: Int, y: Int, abgrColor: Int -> obj.setPixelRGBA(x, y, abgrColor) })
        ),  // Pixel has a lower weight because it has much more placement possibilities
        STAR(
            5,
            listOf(
                DrawFunction { image: NativeImage, x: Int, y: Int, color: Int ->
                    image.setPixelRGBA(x + 1, y, color)
                    image.setPixelRGBA(x, y + 1, color)
                    image.setPixelRGBA(x - 1, y, color)
                    image.setPixelRGBA(x, y - 1, color)
                }
            )),
        THINGY(
            2,
            listOf(
                DrawFunction { pixels: NativeImage, x: Int, y: Int, col: Int ->
                    pixels.setPixelRGBA(x + 1, y, col)
                    pixels.setPixelRGBA(x, y + 1, col)
                    pixels.setPixelRGBA(x - 1, y, col)
                    pixels.setPixelRGBA(x, y - 1, col)

                    val transCol = FastColor.ARGB32.multiply(col, -0x1666667)
                    pixels.setPixelRGBA(x + 1, y + 1, transCol)
                    pixels.setPixelRGBA(x - 1, y - 1, transCol)
                }
            )),
        STAR_ANIMATED(
            3,
            listOf(
                DrawFunction { image: NativeImage, x: Int, y: Int, color: Int ->
                    image.setPixelRGBA(x + 1, y, color)
                    image.setPixelRGBA(x, y + 1, color)
                    image.setPixelRGBA(x - 1, y, color)
                    image.setPixelRGBA(x, y - 1, color)
                },
                DrawFunction { image: NativeImage, x: Int, y: Int, color: Int ->
                    image.setPixelRGBA(x + 1, y, color)
                    image.setPixelRGBA(x, y + 1, color)
                    image.setPixelRGBA(x - 1, y, color)
                    image.setPixelRGBA(x, y - 1, color)
                    image.setPixelRGBA(x + 2, y, color)
                    image.setPixelRGBA(x - 2, y, color)
                    image.setPixelRGBA(x, y + 2, color)
                    image.setPixelRGBA(x, y - 2, color)

                    val transCol = FastColor.ARGB32.multiply(color, -0x1666667)
                    image.setPixelRGBA(x + 1, y + 1, transCol)
                    image.setPixelRGBA(x - 1, y - 1, transCol)
                    image.setPixelRGBA(x - 1, y + 1, transCol)
                    image.setPixelRGBA(x + 1, y - 1, transCol)
                }
            ));

        @JvmField
        val weight: Weight = Weight.of(weight)

        fun draw(image: NativeImage, x: Int, y: Int, color: Color, frame: Int) {
            // Format: ARGB
            frames[frame].draw(image, x, y, color.toInt())
        }

        fun frameCount(): Int {
            return frames.size
        }

        fun hasAnimation(): Boolean {
            return frames.size > 1
        }

        override fun getWeight(): Weight {
            return this.weight
        }

        fun interface DrawFunction {
            fun draw(image: NativeImage, x: Int, y: Int, color: Int)
        }

        companion object {
            private val weightedRandomList: WeightedRandomList<Style> = WeightedRandomList.create(*entries.toTypedArray())

            fun weighted(rng: RandomSource): Style {
                return weightedRandomList.getRandom(rng).get()
            }
        }
    }

    companion object {
        val TRANSPARENCY: SimpleWeightedRandomList<Int> =
            SimpleWeightedRandomList.Builder<Int>().add(0, 5).add(1, 2).add(2, 1).build()
    }
}